debezium-embedded/src/main/java/io/debezium/embedded/AsyncEmbeddedEngine.java

```diff
@@ -0,0 +1,1266 @@
+            return this;
+        }
+
+        @Override
+        public Builder<SourceRecord> using(final OffsetCommitPolicy policy) {
+            this.offsetCommitPolicy = policy;
+            return this;
+        }
+
+        @Override
+        public DebeziumEngine<SourceRecord> build() {
+            return new AsyncEmbeddedEngine(config, consumer, handler, classLoader, clock, completionCallback, connectorCallback, offsetCommitPolicy, null);
+        }
+    }
+
+    /**
+     * Build the default {@link DebeziumEngine.ChangeConsumer} from provided custom {@link Consumer}.
+     *
+     * @param consumer {@link Consumer} provided by the user.
+     * @return {@link DebeziumEngine.ChangeConsumer} which use user-provided {@link Consumer} for processing the Debezium records.
+     */
+    private static ChangeConsumer<SourceRecord> buildDefaultChangeConsumer(Consumer<SourceRecord> consumer) {
+        return new DebeziumEngine.ChangeConsumer<>() {
+
+            /**
+             * The default implementation of {@link DebeziumEngine.ChangeConsumer}.
+             * On every record, it calls the consumer, and then only marks the record
+             * as processed when accept returns. Additionally, it handles StopEngineException
+             * and ensures that we always try and mark a batch as finished, even with exceptions.
+             *
+             * @param records the records to be processed
+             * @param committer the committer that indicates to the system that we are finished
+             *
+             * @throws Exception
+             */
+            @Override
+            public void handleBatch(final List<SourceRecord> records, final DebeziumEngine.RecordCommitter<SourceRecord> committer) throws InterruptedException {
+                for (SourceRecord record : records) {
+                    try {
+                        consumer.accept(record);
+                        committer.markProcessed(record);
+                    }
+                    catch (StopEngineException ex) {
+                        // Ensure that we mark the record as finished in this case.
+                        committer.markProcessed(record);
+                        throw ex;
+                    }
+                }
+                committer.markBatchFinished();
+            }
+        };
+    }
+
+    /**
+     * Possible engine states.
+     * Engine state must be changed only via {@link AsyncEmbeddedEngine#setEngineState(State, State)} method.
+     */
+    private enum State {
+        // Order of the possible states is important, enum ordinal is used for state comparison.
+        CREATING, // the engine is being started, which mostly means engine object is being created or was already created, but run() method wasn't called yet
+        INITIALIZING, // initializing the connector
+        CREATING_TASKS, // creating connector tasks
+        STARING_TASKS, // starting connector tasks
```

Comment:
```suggestion
        STARTING_TASKS, // starting connector tasks
```
it is even in the spec :-)

Reply:
I have no particular intention wrt naming. I'm fine with whatever sounds better or whatever @Naros suggests.

As for the start/stop connectors: IIRC @jpechane explicitly wanted that (at least for now) the engine should manage exactly one connector, so start/stop of the connector should be equivalent to start/stop whole engine. In the future we may reconsider this (especially running source connector and jdbc sink connector withing one engine would make a lot of sense to me), but it's not part of this work.

Reply:
> I only want to make sure this is intentional, because on the first read I felt like names should be TASKS_CREATING, TASKS_STARTING, TASKS_POLLING. On the second look I did not feel like this has to be changed but could, but I wanted to ask your opinion

So this really generally boils down to a question of what is the focus of this enum.  I'd say that our focus is on the state the engine is in, "starting" / "initializing" / "running" / "polling", and less so on the intermediate object that is being acted upon.  This also aligns with what Vojta chose.  If our focus was more on the object being acted upon and less so on the action being taken, then using "<noun>_<verb>" form would be more appropriate vernacular.

Reply:
Thanks for the discussion. I must admit I don't have any preference to it either to I cowardly defer to the native speaker too :-). I beleive it can go as is now so I am merging the PR as the names can be change later.